### ISSUE TITLE
net/mail: mail.ParseAddress() accepts address with domain part starting with a dash

### ISSUE THREAD

**Issue by bmassemin** [golang/go#39488]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.13 darwin/amd64
go go1.14.3 playground
</pre>

### Does this issue reproduce with the latest release?

Yes: https://play.golang.org/p/Z3a5zj6Qch6

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/bmassemin/Library/Caches/go-build"
GOENV="/Users/bmassemin/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/bmassemin"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.13/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.13/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/tn/b45tkwwd0dg36r6yk1bzvzv40000gn/T/go-build066110340=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

https://play.golang.org/p/Z3a5zj6Qch6

```
package main

import (
	"net/mail"
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test(t *testing.T) {
	_, err := mail.ParseAddress("test@--domain.com")
	assert.Error(t, err)
	_, err = mail.ParseAddress("test@domain--.com")
	assert.Error(t, err)
}
```

### What did you expect to see?

I'm expecting the test to pass, since `test@--domain.com` is not a valid address. 
A domain can't starts or ends with a dash according to this [RFC](https://tools.ietf.org/html/rfc1035#section-2.3.1) 
### What did you see instead?

The email address is properly parsed with the invalid domain.

**Comment by odeke-em** [comment]
Thank you for reporting this @bmassemin with detail, and welcome to the Go project!

Those addresses are valid according to the specification that net/mail follows aka [RFC 5322](https://tools.ietf.org/html/rfc5322#page-16), which is loose of a specification that what you are referencing in [RFC 1035](https://tools.ietf.org/html/rfc1035#section-2.3.1) and correct according to RFC 5322's syntax in which it is accepted as a dot-atom per [atom](https://tools.ietf.org/html/rfc5322#section-3.2.3)
![Screen Shot 2020-06-09 at 3 44 28 PM](https://user-images.githubusercontent.com/4898263/84208244-1f7fa080-aa68-11ea-967f-1364ae7827b6.png)

![Screen Shot 2020-06-09 at 3 41 36 PM](https://user-images.githubusercontent.com/4898263/84208101-d0397000-aa67-11ea-9726-98731aa6f730.png)

However, in that RFC 5322 that net/mail uses, they say that enforcement of stricter address validations are left to the discretion of the implementers per https://tools.ietf.org/html/rfc5322#section-3.4.1
![Screen Shot 2020-06-09 at 3 43 49 PM](https://user-images.githubusercontent.com/4898263/84208210-0971e000-aa68-11ea-8139-d837ae636c79.png)

This bug will then perhaps a discussion of whether we should tighten address validation to dip into the requests of RFC 1035 et al, or whether we should leave it alone. I shall retitle this bug.

Kindly /cc-ing @iWdGo @dsymonds @ianlancetaylor

**Comment by dsymonds** [comment]
Yes, `net/mail` deliberately sticks closely to the spec, and, if memory serves, actually accepts slightly more. People have very weird old mail boxes that they like to be able to parse, so being a bit more liberal in the parsing was a deliberate choice.

I'd say leave `net/mail` as it is.

**Comment by beoran** [comment]
I agree this can stay as is, however I would still add a note in the comments of ParseAddress to explain which RFC is implemented.

**Comment by dsymonds** [comment]
> I agree this can stay as is, however I would still add a note in the comments of ParseAddress to explain which RFC is implemented.

It already does that in the package doc comment.
